contracts/Tokenator.sol

```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.24;
+
+import {ERC20, ERC20Burnable} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {ITokenator} from "./interfaces/ITokenator.sol";
+
+contract Tokenator is ITokenator, ERC20Burnable, Ownable {
+    using SafeERC20 for IERC20;
+
+    IERC20 public underlyingToken;
+    uint256 private constant PRECISION = 1e18;
+    uint256 public immutable commencement;
+    uint256 public immutable durationSeconds;
+
+    /// @notice the deployer is responsible to send underlying token to the contract.
+    /// @param name_ token name
+    /// @param symbol_ token symbol
+    /// @param totalSupply_ token total supply
+    /// @param underlyingToken_ the token that tokenator can be redeemed for
+    /// @param commencement_ timestamp at which tokenator vesting begins
+    /// @param durationDays_ duration in days for tokenator to fully vest
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        uint256 totalSupply_,
+        IERC20 underlyingToken_,
+        uint256 commencement_,
+        uint256 durationDays_
+    ) ERC20(name_, symbol_) Ownable(msg.sender) {
+        if (commencement_ < block.timestamp) revert CommencementPassed();
+        if (address(underlyingToken_) == address(0)) revert ZeroAddress();
+
+        underlyingToken = underlyingToken_;
+        commencement = commencement_;
+        durationSeconds = durationDays_ * 1 days;
+
+        _mint(msg.sender, totalSupply_);
+
+        emit Deployed(
+            msg.sender,
+            underlyingToken_,
+            commencement_,
+            durationDays_
+        );
+    }
+
+    /// @notice allows user to claim underlying token from this contract by burning tokenator token
+    /// @param tokenatorAmount amount of tokenator token to burn
+    /// @param to address to receive the underlying token
+    /// @return underlyingTokenAmount amount of underlyingToken `to` address receives
+    function claimUnderlyingToken(
+        uint256 tokenatorAmount,
+        address to
+    ) external returns (uint256 underlyingTokenAmount) {
```

Comment:
let's add non-reentrancy guard here (and to rescue/emergency withdraw

Reply:
why? there's no re-entrancy risk in any of these functions

Reply:
underlyingToken is known, so should be no risk in  claimUnderlyingToken and emergencyWithdraw. But tokens re unknown in rescueTokens, so I think re-entrance guard could be used there. But if we use it in one place, we could just as well have it on the other two. I don't feel strongly about it, but would rather overprotect. Gas is not an issue here

Reply:
true but in `rescueTokens` there are no state changes happening after the external call, so no re-entrancy risk. Plus its an `onlyOwner` function

Reply:
ok